hnn_core/network.py

```diff
@@ -219,47 +286,412 @@ def __repr__(self):
+        seedcore : int
+            Optional initial seed for random number generator (default: 42).
+            Each artificial drive cell has seed = seedcore + gid
         """
-        # params of common external feeds inputs in p_common
-        # Global number of external inputs ... automatic counting
-        # makes more sense
-        # p_unique represent ext inputs that are going to go to each cell
-        self._p_common, self._p_unique = create_pext(self.params,
-                                                     self.params['tstop'])
-        self._n_common_feeds = len(self._p_common)
-
-        # 'position' the artificial cells arbitrarily in the origin of the
-        # network grid. Non-biophysical cell placement is irrelevant
-        # However, they must be present to be included in gid_ranges!
-        origin = self.pos_dict['origin']
-
-        # COMMON FEEDS
-        self.pos_dict['common'] = [origin for i in range(self._n_common_feeds)]
-
-        # UNIQUE FEEDS
-        for key in self._p_unique.keys():
-            # create the pos_dict for all the sources
-            self.pos_dict[key] = [origin for i in range(self.n_cells)]
-
-        # Now add the names of the feeds to a list
-        self.feedname_list.append('common')
-        # grab the keys for the unique set of inputs and sort the names
-        # append them to the src list along with the number of cells
-        unique_keys = sorted(self._p_unique.keys())
-        self.feedname_list += unique_keys
+        if T < 0.:
+            raise ValueError('End time of Poisson input cannot be negative')
+        tstop = self.cell_response.times[-1]
+        if T > tstop:
+            raise ValueError(f'End time of Poisson drive cannot exceed '
+                             f'simulation end time {tstop}. Got {T}.')
+        duration = T - t0
+        if duration < 0.:
+            raise ValueError('Duration of Poisson drive cannot be negative')
+
+        if not isinstance(rate_constants, dict):
+            raise ValueError('rate_constants must be a dict of floats')
+        if not set(rate_constants.keys()).issubset(self.cellname_list):
+            raise ValueError(
+                f"Rate constants should be defined for {self.cellname_list}. "
+                f"Got {list(rate_constants.keys())}")
+
+        drive = NetworkDrive()
+        drive['type'] = 'poisson'
+        drive['cell_specific'] = True
+        drive['seedcore'] = seedcore
+
+        drive['dynamics'] = dict(t0=t0, T=T, rate_constants=rate_constants)
+        drive['events'] = []
+        self._attach_drive(name, drive, weights_ampa, weights_nmda, location,
+                           space_constant, dispersion_time)
+
+    def add_bursty_drive(self, name, distribution, t0, sigma_t0, T, burst_f,
+                         burst_sigma_f, numspikes, repeats, weights_ampa,
+                         weights_nmda, location, dispersion_time=0.1,
+                         space_constant=100., seedcore=42):
```

Comment:
should we force users to pass named arguments:

https://stackoverflow.com/questions/2965271/forced-naming-of-parameters-in-python

? I think I saw sklearn do something similar recently + it will give us flexibility to change the order of the parameters without breaking user scripts.

Reply:
Agreed. Should we do this though (note the asterisk)?

```python
def add_bursty_drive(self, name, *, distribution=, ...)
```

Reply:
yep, that sounds good!